### ISSUE TITLE
ES2015 destructiring assignment with renaming causes key-spacing

### ISSUE THREAD

**Issue by davidmaxwaterman** [eslint/eslint#8451]
I'm seeing this same error for reassignment during destructuring, as done here :

https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Assigning_to_new_variable_names

I accept that it might be considered appropriate to add a space after the colon, as in the case of a key/value pair, but the wording of the error is a little misleading since it isn't a key/value pair. Is it worth rewording the error message, or at least adding such an example to the docs so people can see it is not a mistake?

**Tell us about your environment**

* **ESLint Version:**
v3.19.0
* **Node Version:**
v7.8.0
* **npm Version:**
4.2.0

**What parser (default, Babel-ESLint, etc.) are you using?**
?

**Please show your full configuration:**

```
{
  "extends": ["eslint:recommended", "google"],
  "env": {
    "browser": true,
    "es6": true
  },
  "plugins": [
    "html"
  ],
  "rules": {
    "no-var": "off",
    "new-cap": ["error", { "capIsNewExceptions": ["Polymer", "PolymerRedux"] }],
    "no-console": ["error", { "allow": ["info", "warn", "error"] }]
  },
  "globals": {
    "Polymer": true
  },
  "parserOptions": {
    "ecmaVersion": 6
  }
}
```

**What did you do? Please include the actual source code causing the issue.**

```js
let {b:c} = {a: 1};
```

**What did you expect to happen?**
Nothing: ie is it acceptable for there to be no space after the ':' when renaming during destructuring,
or:
A relevant, helpful error message, such as 'Missing space before assigned variable in destructuring' - I'm sure there is something better than that.

**What actually happened? Please include the actual, raw output from ESLint.**

```
  1:8  error  Missing space before value for key 'b'  key-spacing
  1:8  error  'c' is assigned a value but never used  no-unused-vars

✖ 2 problems (2 errors, 0 warnings)
```

The current message is un-intuitive (imo) since it references a key-value pair, and this isn't that.

https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Assigning_to_new_variable_names

Tagging @btmills, as requested.

**Comment by btmills** [comment]
Thanks for filing this. I was able to reproduce it minimally in the [demo](http://eslint.org/demo):

```js
/* eslint key-spacing: 2 */
/* eslint-disable no-undef, no-unused-vars */
({b:c} = {a: 1}) // 3:5 - Missing space before value for key 'b'. (key-spacing)
```

We have to be careful what we change this message to because there could be patterns other than an identifier on the right side of the colon:

```js
/* eslint key-spacing: 2 */
/* eslint-disable no-undef, no-unused-vars */
({e:{f}} = {d: 1})
```

@davidmaxwaterman what about "Missing space after colon for destructured key 'e'"? If we decide to make a change here, would you be willing to implement it?

**Comment by borela** [comment]
@btmills Is it possible to disallow spacing for aliases but require it for key value pairs?

**Comment by btmills** [comment]
@borela no, they're treated the same currently

**Comment by chipit24** [comment]
Any update on this? I would like to be able to enforce a structure like this:
```
const myObj = { id: '1234' };
const { id:ObjId } = myObj;
```
I think disallowing the space after the colon for the renamed variable makes it more clear that it's a renamed variable and not a value in a key-value pair.

**Comment by btmills** [comment]
@chipit24 The configuration for `key-spacing` is already pretty complex, so I suspect the documentation change has a better chance of happening, and even that hasn't gotten much interest

**Comment by davidmaxwaterman** [comment]
Aside: what is this: `({e:{f}} = {d: 1})`?

**Comment by btmills** [comment]
@davidmaxwaterman here's a less contrived example:

```js
// This:
const {
    location: {
        protocol
    }
} = window;

// Is equivalent to this:
const protocol = window.location.protocol;
```

Check out [nested object destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Nested_object_and_array_destructuring) on MDN.

**Comment by borela** [comment]
In case anybody is using sublime text, [I solved the issue by adding specific rules to the syntax highlighting Naomi in combination with the color scheme Candyman](https://github.com/borela/naomi) to make it easier to distinguish when variables were renamed.
![untitled](https://user-images.githubusercontent.com/11317458/30786134-b5ffac08-a147-11e7-9a08-c46463bdf357.png)

**Comment by anthonator** [comment]
I know this isn't likely to happen due to the complexity of the `key-spacing` rules but I'll throw my hat into the "I'd like to ignore `key-spacing` when destructuring" ring.

If someone were willing to put in the effort would the `eslint` team be willing to accept an option to treat destructuring differently? If so, how do you think that would look? An `ignoreDestructuring` boolean option? Something more complex?
